packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -32,84 +32,123 @@ export interface CompletionProvider {
-                stream(features).distinct(e => {
-                    if (ast.isKeyword(e)) {
-                        return e.value;
-                    } else {
-                        return e;
-                    }
-                }).forEach(e => this.completionFor(findRelevantNode(node), e, acceptor));
-            } else {
-                // The entry rule is the first parser rule
-                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-                const parserRule = this.grammar.rules.find(e => ast.isParserRule(e))!;
-                this.completionForRule(undefined, parserRule, acceptor);
+            };
+
+            stream(beforeFeatures)
+                .distinct(distinctionFunction)
+                .forEach(e => this.completionFor(findRelevantNode(node), e, acceptor));
+
+            if (reparse) {
+                const missingPart = document.textDocument.getText({
+                    start: document.textDocument.positionAt(parserStart),
+                    end: params.position
+                }).toLowerCase();
+                // Remove items from `beforeFeatures` which don't fit the current text
+                items = items.filter(e => e.label.toLowerCase().startsWith(missingPart));
+
+                stream(afterFeatures)
+                    .exclude(beforeFeatures, distinctionFunction)
+                    .distinct(distinctionFunction)
+                    .forEach(e => this.completionFor(findRelevantNode(node), e, acceptor));
             }
         }
         return CompletionList.create(items, true);
     }
 
-    protected buildFeatureStack(node: CstNode | undefined): ast.AbstractElement[] {
-        const features: ast.AbstractElement[] = [];
-        while (node) {
-            if (node.feature) {
-                features.push(node.feature);
-            }
-            node = node.parent;
+    protected findFeaturesAt(document: TextDocument, offset: number): ast.AbstractElement[] | false {
+        const text = document.getText({
+            start: Position.create(0, 0),
+            end: document.positionAt(offset)
+        });
+        const parserResult = this.completionParser.parse(text);
+        // If the parser didn't parse any tokens, just return false
+        if (!parserResult.token) {
+            return false;
         }
+        const leftoverTokens = [...parserResult.tokens].splice(parserResult.tokenIndex);
+        const features = findNextFeatures(parserResult.elementStack, leftoverTokens);
         return features;
     }
 
+    protected backtrackToTokenStart(text: string, offset: number): number {
+        const wordRegex = /\w/;
+        const whiteSpaceRegex = /\s/;
+        let lastCharacter = text.charAt(offset - 1);
+        if (whiteSpaceRegex.test(lastCharacter)) {
+            return offset;
+        }
+        while (wordRegex.test(lastCharacter) && offset > 0) {
+            offset--;
+            lastCharacter = text.charAt(offset - 1);
```

Comment:
Wouldn't this cause issues if `offset` is ever `1`. This would end up retrieving a char from `-1` in such a case.